### ISSUE TITLE
Towards release v2020.08.07

### ISSUE THREAD

**Issue by tahina-pro** [project-everest/everparse#24]
This is a meta-issue for the following improvements:

- [x] Given a module `A.3d`, generate a separate `A.Types.fst` to make arith error reporting more user-friendly (#23 )
- [x] Check that a `.3d`  file has at least one `entrypoint` (#20 )
- [x] Further improve KReMLin warnings 7, 8 wrt. `static inline`
- [x] Allow generating C++-compatible headers wrapped into `extern "C"` ~(not needed, since it is enough for the caller to wrap their `#include`s into an `extern "C"` block)~(that solution is not robust wrt. transitive `#include`s)
- ~Allow `struct` arguments, desugar them into their fields~ Not for now, see https://github.com/project-everest/everparse/issues/24#issuecomment-670249294
- [x] Modulus as an arithmetic operator
- [x] Bitwise operators on integer types
- [x] Multi-line comments (currently only allowed at specific points in AST and propagated to extracted code)
- [x] Allow additional F* and KReMLin arguments to 3d via environment variables (#22 for KReMLin)

**Comment by tahina-pro** [comment]
For the record, re. KReMLin warnings:
* spurious warning 8 now avoided thanks to FStarLang/KReMLin@7d0459dfdfcb844977452f9be6055458ae91f97c
* spurious warning 7 now avoided thanks to 61aaf9c46a3a63c22ff094ab55870815030eab56 sqq.

**Comment by tahina-pro** [comment]
`extern "C"` can now be added to .h files via KReMLin's `-fextern-c` option, cf. FStarLang/kremlin@5940bfa882e02c5431aaaf375beba03a8fb75ea5
With #22 I should be able to allow users to pass this option to 3d

**Comment by tahina-pro** [comment]
unless you prefer to have a specific 3d command-line option for that. Or make it as default. @msprotz @nikswamy  What do you think?

**Comment by msprotz** [comment]
I would say default is good -- it doesn't hurt in C, doesn't hurt in C++ either. I just wanted some wider testing before I made it the default itself in KreMLin.

**Comment by tahina-pro** [comment]
ok, I now call KReMLin with `-fextern-c` by default.

**Comment by tahina-pro** [comment]
Users can now set the `EVERPARSE_FSTAR_OPTIONS` (resp. `EVERPARSE_KREMLIN_OPTIONS`) environment variable to provide their own additional arguments to F* (resp. KReMLin), see ab9f27376bcdacb6e87645f5fa01daab6d46b43f

**Comment by nikswamy** [comment]
After a chat with @tahina-pro, we decided that we were not going to do struct passing, at least not for this release.

It is more complex that we expected, for a couple of reasons.

1. If the user write a struct as an argument in a 3d spec, then we would explode it into its fields. But, those struct types in a 3d definition could have refinements associated with each field. So, to preserve those refinements we would have to explode the struct into its simply typed fields and then move any refinements on all fields and collect them in a where clause of the struct. This is doable, but non-trivial.

2. More problematic is the API to provide to the validator in the generated wrapper. It should presumably expose a signature allowing the caller to pass in a struct, rather than all the exploded fields of the struct. But what type should it use for that struct? That type is defined presumably in some caller defined header ... but the 3d file has no awareness of that header. Perhaps this could be better done from c3d, where the types are being generated from an existing C header.

In summary, more complex that we thought: not sure it's worth it. Perhaps we'll address it in a future release, if there is strong demand for it.

**Comment by tahina-pro** [comment]
I just released v2020.08.07: https://github.com/project-everest/everparse/releases/tag/v2020.08.07
Thank you everyone (@aseemr , @nikswamy , @msprotz , in no particular order) for having contributed to this success!
